### DISCUSSION TITLE
mssql+pyodbc dialect does not have a configurable port option

### DISCUSSION THREAD

**Discussion by dxdc** [sqlalchemy/sqlalchemy#8271] [Usage Questions] [answered]
### Describe the bug

`mssql+pyodbc` dialect does not respect the port option. Perhaps a similar fix is needed for this dialect as here:

https://github.com/sqlalchemy/sqlalchemy/commit/a29f130be7a1589566547e94c0e283f0210f27ef

### To Reproduce

```python
Create any engine string using this dialect, and supply a different port number than the default (e.g., 1435). It will still connect to port 1433.
```

### Error

Still connects to port 1433, regardless of port supplied in engine config.

### Versions

- OS: Mac
- Python: 3.9
- SQLAlchemy: latest
- Database: mssql
- DBAPI (eg: psycopg, cx_oracle, mysqlclient): mssql+pyodbc

**Comment by zzzeek** [comment, marked as answer]
hiya -

the port number is used for [pyodbc hostname connections](https://github.com/mkleehammer/pyodbc/wiki/Connecting-to-SQL-Server-from-Linux) only, that is:


```py
$ python
Python 3.10.0 (default, Nov  5 2021, 17:23:47) [GCC 11.2.1 20210728 (Red Hat 11.2.1-1)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from sqlalchemy.engine import url
>>> u = url.make_url("mssql+pyodbc://scott:tiger@somehost:1435/somedb?driver=ODBC+Driver+17+for+SQL+Server")
>>> from sqlalchemy.dialects.mssql import pyodbc
>>> pyodbc.dialect().create_connect_args(u)
[['DRIVER={ODBC Driver 17 for SQL Server};Server=somehost,1435;Database=somedb;UID=scott;PWD=tiger'], {}]

```

above we can see "Server=somehost,1435", there's the port number.  That syntax is the [documented pyodbc syntax for linux hostname connections](https://github.com/mkleehammer/pyodbc/wiki/Connecting-to-SQL-Server-from-Linux).

You're using OSX, which seems to have [a totally different API](https://github.com/mkleehammer/pyodbc/wiki/Connecting-to-SQL-Server-from-Mac-OSX) for a DSN-less connection (it's also not fully clear if this is config-less or not as the docs are ambiguous if you still need an entry in odbcinst.ini).

as OSX seems to document that you have to use freetds and not a microsoft driver, things are going to be less friendly here and I'd advise you set up a regular DSN in odbc.ini, I used freetds for many years and it is not very good.

**Reply by zzzeek** [reply]
@gordthompson do you know anything about string URL formats for pyodbc + osx ?

**Reply by dxdc** [reply]
thanks for the reply @zzzeek. sorry, I should have indicated, I'm running a Mac (OSX), but running this via a Docker container that is running Debian GNU/Linux 10 (buster).


```
# install SQL Server Driver 18
RUN apt-get install -y msodbcsql18
```

I am generating the engine URL as follows:

```py
from sqlalchemy.engine import URL
from sqlalchemy.dialects.mssql import pyodbc

query = dict(
	driver="ODBC Driver 18 for SQL Server",
	TrustServerCertificate="Yes"
)

engine_url = URL.create(
	drivername="mssql+pyodbc",
	username="user",
	password="pw",
	port="1345",
	host="localhost",
	database="db",
	query=query
)

print(engine_url)
# mssql+pyodbc://user:pw@localhost:1345/db?TrustServerCertificate=Yes&driver=ODBC+Driver+18+for+SQL+Server

print(pyodbc.dialect().create_connect_args(engine_url))
# [['DRIVER={ODBC Driver 18 for SQL Server};Server=localhost,1345;Database=db;UID=user;PWD=pw;TrustServerCertificate=Yes'], {}]
```

I don't see anything obviously wrong with the engine string? But, regardless of this string and the port value, it still tries to connect to port 1433.

**Reply by zzzeek** [reply]
when I try it here running the test suite, changing the port causes it to no longer connect and it times out.    try connecting with pyodbc directly using that string.

**Reply by dxdc** [reply]
thx @zzzeek! I figured it out, it was another configuration issue conflicting on my side. It's not a bug, please disregard. Thanks for sharing the `pyodbc.dialect().create_connect_args()` function, I haven't seen that before!

**Comment by kshpytsya** [comment]
For what it is worth, `localhost` didn't work for me but `127.0.0.1` did. For `localhost` case strace showed repeated dns queries when connection to server was expected.

**Reply by CaselIT** [reply]
That seems like an issue with your environment / driver, it's not something that's controlled by sqlalchemy.
Makes sense mentioning it though
